src/clj/coffi/mem.clj

```diff
@@ -1350,3 +1766,523 @@
+            (map-seq       [] (list 'seq         ['this]           (list `StructMapSeq. 'this 0)))
+            (invoke1       [] (list 'invoke      ['this 'arg1]       (concat [`case 'arg1] (interleave (range) as-vec) (interleave members as-vec) [nil])))
+            (invoke2       [] (list 'invoke      ['this 'arg1 'arg2] (concat [`case 'arg1] (interleave (range) as-vec) (interleave members as-vec) ['arg2])))
+            (applyTo       [] (list 'applyTo      ['this 'arglist] (concat [`case (list `first 'arglist)] (interleave (range) as-vec) (interleave members as-vec) [(list `if (list `.next 'arglist) (list `.first (list `.next 'arglist)) nil)])))
+            ;structimpl utility function
+            (s-nth-key     [] (list 'nthKey     ['this 'i]        (concat [`case 'i] (interleave (range) members))))
+            ;java.util.Map implementations
+            (map-contains-value [] (list 'containsValue ['this 'val] (list `some (set as-vec) 'val)))
+            (map-entrySet       [] (list 'entrySet ['this] (set (map (fn [[k v]] (list `clojure.lang.MapEntry/create k v)) (partition 2 (interleave members as-vec))))))
+            (map-get            [] (cons 'get (rest (s-valAt))))
+            (map-isEmpty        [] (list 'isEmpty ['this] false))
+            (map-keySet         [] (list 'keySet ['this] (set members)))
+            (map-size           [] (list 'size ['this] (count members)))
+            (map-values         [] (list 'values ['this] as-vec))
+
+            (map-methods   [] [(map-without) (map-cons) (map-equiv) (map-empty) (map-iterator) (map-foreachConsumer) #_(map-foreachBiConsumer) (map-seq) (map-assoc) (map-assocEx) (map-contains-value) (map-entrySet) (map-get) (map-isEmpty) (map-keySet) (map-size) (map-values)])
+            (vec-methods   [] [(vec-length) (vec-assoc) (vec-assocN) (vec-peek) (vec-pop) (vec-nth) (vec-nth-2) (vec-cons) (vec-equiv) (vec-empty) (vec-iterator) (vec-foreach) (vec-seq) (vec-rseq)])
+            (struct-methods [] [(s-count) (s-containsKey) (s-valAt) (s-valAt-2) (s-entryAt)])
+            (prefix-methods [prefix ms] (map (fn [[method-name & tail]] (cons (with-meta (symbol (str prefix method-name)) (meta method-name)) tail)) ms))
+            (impl-methods [] (concat (prefix-methods "map_" (map-methods)) (prefix-methods "vec_" (vec-methods)) (prefix-methods "struct_" (struct-methods))))]
+      (concat
+       [`deftype (symbol (name typename)) (vec typed-member-symbols) `coffi.mem.IStruct `coffi.mem.IStructImpl `clojure.lang.IPersistentMap `clojure.lang.MapEquivalence `java.util.Map `clojure.lang.IFn]
+       (struct-methods)
+       (map-methods)
+       (impl-methods)
+       [(s-nth-key)
+        (invoke1)
+        (invoke2)
+        (list 'asMap ['this] 'this)
+        (list 'asVec ['this] (list `VecWrap. 'this))]))))
```

Comment:
Not required, but it seems a bit more clear to me when using the normal quasiquote syntax.
```suggestion
      `(deftype ~(symbol (name typename)) ~(vec typed-member-symbols)
         coffi.mem.IStruct
         ~@(struct-methods)
         coffi.mem.IStructImpl
         ~@(impl-methods)
         clojure.lang.IPersistentMap
         clojure.lang.MapEquivalence
         java.util.Map
         ~@(map-methods)
         clojure.lang.IFn
         ~(s-nth-key)
         ~(invoke1)
         ~(invoke2)

         (~'asMap [~'this] ~'this)
         (~'asVec [~'this] (VecWrap. ~'this)))
```

Reply:
I have no strong preference. I think your change looks very readable! I like the grouped Interfaces and then emitted method definitions!